litestar/plugins/problem_details.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Callable, Mapping, TypeVar
+
+from litestar.connection.request import Request
+from litestar.exceptions.http_exceptions import HTTPException
+from litestar.plugins.base import InitPluginProtocol
+from litestar.response.base import Response
+from litestar.serialization.msgspec_hooks import encode_json
+from litestar.types.callable_types import ExceptionHandler, ExceptionT
+
+if TYPE_CHECKING:
+    from litestar.config.app import AppConfig
+
+ProblemDetailsExceptionT = TypeVar("ProblemDetailsExceptionT", bound="ProblemDetailsException")
+ProblemDetailsExceptionHandlerType = Callable[[Request, ProblemDetailsExceptionT], Response]
+ExceptionToProblemDetailMapType = Mapping[type[ExceptionT], Callable[[ExceptionT], ProblemDetailsExceptionT]]
+
+
+def _problem_details_exception_handler(request: Request, exc: ProblemDetailsException) -> Response:
+    return exc.to_response(request)
+
+
+def _create_exception_handler(
+    exc_to_problem_details_exc_fn: Callable[[ExceptionT], ProblemDetailsException], exc_type: type[ExceptionT]
+) -> ExceptionHandler:
+    def _exception_handler(req: Request, exc: exc_type) -> Response:
+        problem_details_exc = exc_to_problem_details_exc_fn(exc)
+
+        return problem_details_exc.to_response(req)
+
+    return _exception_handler
+
+
+class ProblemDetailsException(HTTPException):
+    _PROBLEM_DETAIL_HEADER = "application/problem+json"
+
+    def __init__(
+        self,
+        type_: str = "about:blank",
+        status_code: int | None = None,
+        title: str | None = None,
+        instance: str | None = None,
+        detail: str | dict[str, Any] | list[Any] | None = None,
+        extra: dict[str, Any] | None = None,
+        headers: dict[str, Any] | None = None,
+    ) -> None:
+        self.type_ = type_
+        self.status_code = status_code or self.status_code
+        self.title = title
+        self.instance = instance
+        self.detail = detail  # type: ignore
+        self.extra = extra
+        self.headers = headers
+
+    def to_response(self, request: Request) -> Response:
+        problem_details = {"type": self.type_, "status": self.status_code}
+        if self.title is not None:
+            problem_details["title"] = self.title
+        if self.instance is not None:
+            problem_details["instance"] = self.instance
+        if self.detail is not None:
+            problem_details["detail"] = self.detail
+        if self.extra is not None:
+            problem_details.update(self.extra)
+
+        content = encode_json(problem_details, request.route_handler.default_serializer)
+        headers = self.headers or {}
+        headers["content-type"] = self._PROBLEM_DETAIL_HEADER
```

Comment:
Wouldn't it be enough to just have the media type of the underlying HTTPException be application/problem+json? I am a bit surprised to see this setting of content-type and encoding inside the new exception class, since there is already code for all of that.

Reply:
This is part of the (current) RFC spec I believe and the main purpose of this plugin so the overwritten / forced content-type  makes sense to me.

Is there a better place to do the content-type assignment?

Reply:
a `Response` with `media_type="application/problem+json"` will be automatically json-encoded and get the content-type header `"application/problem+json"`.

Reply:
@bunny-therapist you're absolutely right. I forgot about that PR that added support for this. We should just be setting the `media_type` and allow the serialization to happen as per the normal flow.